### PR TITLE
Allow TCPSocket to bind to specific local address

### PR DESCRIPTION
**Pull request by maxpowa**
Allows a TCPSocket to use a specific address for outgoing connections, for use in systems with more than a single interface. Adds two TCPSocket overloads to the existing instantiation methods, `TCPSocket.open` and `TCPSocket.new`, with additional parameters to allow specifying an IP and port combination to bind. Not sure how great the specs are, UDPSocket has much better testing for this but unfortunately TCPSocket cannot be instantiated in the same way as UDPSocket.

cc @RX14

### CODE REVIEW

File: src/socket/tcp_socket.cr [hunk 1]
```diff
@@ -39,6 +68,15 @@ class TCPSocket < IPSocket
...
   # Returns the value of the block.
+  def self.open(host, port, local_host, local_port)
+    sock = new(host, port, local_host: local_host, local_port: local_port)
+    begin
+      yield sock
+    ensure
+      sock.close
+    end
+  end
+
```
Comment (asterite) [inline src/socket/tcp_socket.cr]: We could actually make this just `def self.open(*args, **nargs)` and then forward the arguments to `new`, so we only need one version of this method.

**Comment by ysbaddaden** [comment]
I'm not sure this is so common that it's really worth adding more arguments to TCPSocket (and complicate it).

I've been considering to refactor Socket to look more like Ruby, where Socket has all the low level C-like methods (bind, listen, ...), so we have access to everything, and where TCPSocket and UDPSocket are limited, but simple, abstractions on top of it.

This way one could:

```crystal
sock = Socket.new(Socket::Protocol::TCP)
sock.bind(local_host, local_port)
sock.connect(host, port)
```

Or by making the connection lazy:

```crystal
sock = TCPSocket.new(host, port)
sock.bind(local_host, local_port)
```

**Comment by maxpowa** [comment]
I agree, Ruby/C style is much better but I wanted to avoid breaking anything with this patch. In order to have an explicit bind call would also require an explicit connect call, which TCPSocket does not currently have (but UDPSocket does :confused:).

**Comment by bararchy** [comment]
@ysbaddaden , @maxpowa 
In Ruby I like the different between the "client" and "server" helper wrappers around the socket class.

client: `TCPSocket.new("127.0.0.1", 8080)`  
server: `TCPServer.new(8080)` (listens on 0.0.0.0 on port 8080) and `TCPServer.new("127.0.0.1", 8080)` (listens on 127.0.0.1 on port 8080)  

Shouldn't we conform to this standard ?

**Pull request closed without merging**